src/Workspaces/Core/Portable/Rename/Renamer.RenameDocumentActionSet.cs

```diff
@@ -0,0 +1,128 @@
+            /// </remarks>
+            public async Task<Solution> UpdateSolutionAsync(Solution solution, ImmutableArray<RenameDocumentAction> actions, CancellationToken cancellationToken)
+            {
+                if (solution is null)
+                {
+                    throw new ArgumentNullException(nameof(solution));
+                }
+
+                if (actions.Any(a => !ApplicableActions.Contains(a)))
+                {
+                    throw new ArgumentException(WorkspacesResources.Cannot_apply_action_that_is_not_in_applicableactions);
+                }
+
+                // Prior to updating the solution it's possible the document id has changed between the time 
+                // the document action info was generated and when the solution update is applied. We
+                // do a best effort to still locate the document if the id has changed.
+                var document = GetDocument(solution);
+
+                // If the document was found in the solution then the current id will be durable across actions
+                // since we own the solution snapshot at this point. 
+                var documentId = document.Id;
+
+                // Make sure that the document name and folders are updated to what we expect them to be
+                solution = solution
+                    .WithDocumentName(documentId, _documentName)
+                    .WithDocumentFolders(documentId, _documentFolders);
+
+                // Apply each action individually. Order should not matter
+                foreach (var action in actions)
+                {
+                    document = solution.GetRequiredDocument(documentId);
+                    solution = await action.GetModifiedSolutionAsync(document, _optionSet, cancellationToken).ConfigureAwait(false);
+                }
+
+                return solution;
+            }
+
+            /// <summary>
+            /// Attempts to find the document in the solution. Tries by documentId first, but 
+            /// that's not always reliable between analysis and application of the rename actions
+            /// </summary>
+            private Document GetDocument(Solution solution)
+            {
+                // DocumentId is the best bet for finding a document,
+                // but it's possible the document was renamed or moved
+                // before actions were applied.
+                if (solution.ContainsDocument(_documentId))
+                {
+                    return solution.GetRequiredDocument(_documentId);
+                }
+
+                var project = solution.GetRequiredProject(_projectId);
+                return project.Documents.FirstOrDefault(d => d.Name == _documentName && d.Folders.SequenceEqual(_documentFolders))
+                    ?? throw new InvalidOperationException(WorkspaceExtensionsResources.The_solution_does_not_contain_the_specified_document);
```

Comment:
the fact that this is a thing seems super fishy to me.  if we produced hte actions against a particular solution, why not just use that solution?

I'm trying to figure out the use case when we'd want to apply actoins to a solution that had drifted away.  esp. one where the documentId was no longer around.  That really seems fishy.

Reply:
Consolidating comments. This concern is captured in https://github.com/dotnet/roslyn/pull/41267#discussion_r410007814